standard/statements.md

```diff
@@ -454,29 +454,27 @@ switch_section
     ;
 
 switch_label
-    : 'case' constant_expression ':'
+    : 'case' pattern case_guard?  ':'
     | 'default' ':'
     ;
+    
+case_guard
+    : 'when' expression
+    ;
 ```
 
-A *switch_statement* consists of the keyword `switch`, followed by a parenthesized expression (called the ***switch expression***), followed by a *switch_block*. The *switch_block* consists of zero or more *switch_section*s, enclosed in braces. Each *switch_section* consists of one or more *switch_label*s followed by a *statement_list* ([§13.3.2](statements.md#1332-statement-lists)).
-
-The ***governing type*** of a `switch` statement is established by the switch expression.
-
-- If the type of the switch expression is `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `bool`, `string`, or an *enum_type*, or if it is the nullable value type corresponding to one of these types, then that is the governing type of the `switch` statement.
-- Otherwise, exactly one user-defined implicit conversion shall exist from the type of the switch expression to one of the following possible governing types: `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `string`, or, a nullable value type corresponding to one of those types.
-- Otherwise, a compile-time error occurs.
+A *switch_statement* consists of the keyword `switch`, followed by a parenthesized expression (called the ***switch expression***), followed by a *switch_block*. The *switch_block* consists of zero or more *switch_section*s, enclosed in braces. Each *switch_section* consists of one or more *switch_label*s followed by a *statement_list* ([§13.3.2](statements.md#1332-statement-lists)). Each *switch_label* containing `case` has an associated pattern (§patterns-new-clause) against which the value of the switch expression is tested. If *case-guard* is present, its expression shall be of type `bool` and that expression is evaluated as an additional condition to be satisfied for the case to be considered satisfied.
 
-The constant expression of each `case` label shall denote a value of a type that is implicitly convertible ([§11.2](conversions.md#112-implicit-conversions)) to the governing type of the `switch` statement. A compile-time error occurs if two or more case labels in the same switch statement specify the same constant value.
+A compile-time error occurs if two or more case labels in the same switch statement specify the same pattern and case guard.
```

Comment:
This is not correct.  The constraint only applies to case labels that are unguarded.

```csharp
int i = 12;
switch (i)
{
    case 1 when "".Length == 0:
        break;
    case 1 when "".Length == 0: // OK
        break;
}
```

Reply:
@gafter Ouch. What is the reason for this being OK? (I can think of some, not sure any are “good” ATM)

Reply:
We don't have a concept of case guards being "the same".  For that matter, we don't have a concept of patterns being "the same".

What we do have is an error if a case is unreachable because it is subsumed by previous (unguarded) cases.